### PR TITLE
UI Widget Tabbing

### PR DESCRIPTION
**Pull request by RatMoleRat**
### Contains

This adds the ability to tab between widgets and activate selected widgets. Also, changes some images so that the widgets selected in this manner are clearly visibly highlighted.

A GIF (sped up to show a variety of widgets) is attached below to see how it works. 

widget tabbing gif

### How to test

Open a menu screen in Terasology. Use the tab key to tab around - shift-tab to rotate through in reverse. Press enter to activate a widget. Use the arrow keys to change the selection in an activated dropdown, to scroll when there is an activated dropdown, or to move up/down when a list widget is selected.

### CODE REVIEW

**Comment by GooeyHub** [comment]
Can one of the admins please verify this patch?

**Comment by Cervator** [comment]
This pull request **introduces 1 alert** when merging d181f46e8661a31e3a0ea35960dc0e0c111046d1 into 24f3487c607b0351fc1127b1c0bcd9e59336bbb9 - [view on LGTM.com](https://lgtm.com/projects/g/MovingBlocks/Terasology/rev/pr-1d797d67d005f06dde9bc22f26850d9bea337e74)

**new alerts:**

* 1 for Implicit narrowing conversion in compound assignment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Cervator** [comment]
This pull request **introduces 1 alert** when merging 0f268fa80f482f5be47d54c8b245befadc21ad44 into 24f3487c607b0351fc1127b1c0bcd9e59336bbb9 - [view on LGTM.com](https://lgtm.com/projects/g/MovingBlocks/Terasology/rev/pr-f4bdd166b8f193aae671935f125b0dfff604a39c)

**new alerts:**

* 1 for Implicit narrowing conversion in compound assignment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by jellysnake** [comment]
@GooeyHub add to whitelist

File: engine/src/main/java/org/terasology/rendering/nui/SortOrderSystem.java [hunk 1]
```diff
@@ -103,17 +108,17 @@ public boolean onRelease(float delta, EntityRef target) {
              }
          };
 
-         if (keys.containsKey(Keyboard.Key.LEFT_SHIFT.getId())) {
```
Comment (jellysnake) [inline engine/src/main/java/org/terasology/rendering/nui/SortOrderSystem.java]: This whole method seems like it should be structured using the event system and the built in binding method but that's a topic for a different PR

File: engine/src/main/java/org/terasology/rendering/nui/SortOrderSystem.java [hunk 2]
```diff
@@ -262,7 +267,20 @@ public static void setUsed(ArrayList<Integer> other) {
...
     }
 
-    public static boolean getModifierPressed() {
-        return modifierPressed;
+    public static boolean containsConsole() {
+        if (enabledWidgets != null) {
+            boolean containsConsole = false;
+            for (CoreScreenLayer layer : enabledWidgets) {
+                if (layer instanceof ConsoleScreen) {
+                    LoggerFactory.getLogger("").info("console open");
```
Comment (jellysnake) [inline engine/src/main/java/org/terasology/rendering/nui/SortOrderSystem.java]: This shouldn't be here

File: engine/src/main/java/org/terasology/rendering/nui/SortOrderSystem.java [hunk 3]
```diff
@@ -262,7 +267,20 @@ public static void setUsed(ArrayList<Integer> other) {
...
+                    containsConsole = true;
```
Comment (jellysnake) [inline engine/src/main/java/org/terasology/rendering/nui/SortOrderSystem.java]: You can short circuit here by simply replacing this like so:
```suggestion
                    return true;
```
and then removing the next return below

File: engine/src/main/java/org/terasology/rendering/nui/TabbingManager.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        focusSetThrough = false;
+        currentNum = 0;
+        maxNum = 0;
+        nextNum = 0;
+        usedNums = new ArrayList<>();
+        widgetsList = new ArrayList<>();
+        initialized = true;
+    }
+
+    public static void changeCurrentNum(boolean increase) {
```
Comment (jellysnake) [inline engine/src/main/java/org/terasology/rendering/nui/TabbingManager.java]: This whole method is confusing and I think you can re-write it much simpler

File: engine/src/main/java/org/terasology/rendering/nui/WidgetWithOrder.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            added = true;
+        } else if (!added) {
+            TabbingManager.addToWidgetsList(this);
+            TabbingManager.addToUsedNums(order);
+            added = true;
+        }
+        return order;
+    }
+    @Override
+    public boolean onKeyEvent(NUIKeyEvent event) {
```
Comment (jellysnake) [inline engine/src/main/java/org/terasology/rendering/nui/WidgetWithOrder.java]: This should really be handled but the individual widgets themselves.
Else this method will need to keep growing as more and more widgets need to implement things here

File: engine/src/main/java/org/terasology/rendering/nui/widgets/UICheckbox.java [hunk 6]
```diff
@@ -15,39 +15,30 @@
...
-     */
-    private List<ActivateEventListener> listeners = Lists.newArrayList();
-
     private InteractionListener interactionListener = new BaseInteractionListener() {
 
         @Override
         public boolean onMouseClick(NUIMouseClickEvent event) {
             if (event.getMouseButton() == MouseInput.MOUSE_LEFT) {
                 active.set(!active.get());
-                activate();
```
Comment (jellysnake) [inline engine/src/main/java/org/terasology/rendering/nui/widgets/UICheckbox.java]: You, uh, broke checkboxes with these changes.
Most likely you forgot to call the new `activateWidget` method. You also override it below so you'll need to call super there

File: engine/src/main/java/org/terasology/rendering/nui/widgets/UIList.java [hunk 7]
```diff
@@ -99,11 +108,17 @@ public void onDraw(Canvas canvas) {
     }
 
     private void updateItemListeners() {
-        while (itemListeners.size() > list.get().size()) {
-            itemListeners.remove(itemListeners.size() - 1);
+        /*
+        for (int i=0; i<listMin; i++) {
```
Comment (jellysnake) [inline engine/src/main/java/org/terasology/rendering/nui/widgets/UIList.java]: Please remove

File: engine/src/main/java/org/terasology/rendering/nui/widgets/UIList.java [hunk 8]
```diff
@@ -53,6 +54,9 @@
     private Binding<List<T>> list = new DefaultBinding<>(new ArrayList<>());
     private ItemRenderer<T> itemRenderer = new ToStringTextRenderer<>();
     private Binding<Boolean> canBeFocus = new DefaultBinding<>(true);
+    private int listMin = -1;
```
Comment (jellysnake) [inline engine/src/main/java/org/terasology/rendering/nui/widgets/UIList.java]: Why is this initialised to `-1` if it's set to `0` on the first frame via `onDraw`

File: engine/src/main/java/org/terasology/rendering/nui/widgets/UILoadBar.java [hunk 9]
```diff
@@ -16,22 +16,22 @@
...
 import org.terasology.rendering.nui.ScaleMode;
+import org.terasology.rendering.nui.WidgetWithOrder;
 import org.terasology.rendering.nui.databinding.Binding;
 import org.terasology.rendering.nui.databinding.DefaultBinding;
 
 /**
  * An animated loading bar
  */
-public class UILoadBar extends CoreWidget {
+public class UILoadBar extends WidgetWithOrder {
```
Comment (jellysnake) [inline engine/src/main/java/org/terasology/rendering/nui/widgets/UILoadBar.java]: Why is this a `WidgetWithOrder`?

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by Cervator** [comment]
This pull request **introduces 3 alerts** when merging 3de3d38a74bb596f09e047517b605a772b1ff9a8 into 24f3487c607b0351fc1127b1c0bcd9e59336bbb9 - [view on LGTM.com](https://lgtm.com/projects/g/MovingBlocks/Terasology/rev/pr-760b422d86d0b16f63c426efec6ba270011e19ea)

**new alerts:**

* 2 for Result of multiplication cast to wider type
* 1 for Implicit narrowing conversion in compound assignment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by Cervator** [comment]
This pull request **introduces 3 alerts** when merging 315ae9c24b249663f26063d6bcfd9b7b9274c77d into 24f3487c607b0351fc1127b1c0bcd9e59336bbb9 - [view on LGTM.com](https://lgtm.com/projects/g/MovingBlocks/Terasology/rev/pr-37ec344de8e9dcf5ad61abc9f5399a15c2749b5d)

**new alerts:**

* 2 for Result of multiplication cast to wider type
* 1 for Implicit narrowing conversion in compound assignment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Cervator** [comment]
This pull request **introduces 1 alert** when merging 9734604fb82eba3c6ae28b5a0d98f03b11e5332a into 948676050a7827dac5e04927087832ffc462da41 - [view on LGTM.com](https://lgtm.com/projects/g/MovingBlocks/Terasology/rev/pr-7207b97a71a6abce9285d98ef24a714b30093ce6)

**new alerts:**

* 1 for Implicit narrowing conversion in compound assignment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by Cervator** [comment]
This pull request **introduces 1 alert** when merging e603dfa498d2a80aaec0f6ed1d99d988d71aa18b into 948676050a7827dac5e04927087832ffc462da41 - [view on LGTM.com](https://lgtm.com/projects/g/MovingBlocks/Terasology/rev/pr-4883d72d83fdbdab0215d7f6a6efbfdb44aa8af5)

**new alerts:**

* 1 for Implicit narrowing conversion in compound assignment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/rendering/nui/widgets/UIScrollbar.java [hunk 10]
```diff
@@ -250,5 +250,4 @@ private void updatePosition(int pixelPos) {
         int newPosition = TeraMath.clamp(pixelPos, 0, sliderSize);
         setValue((sliderSize > 0) ? (newPosition * getRange() / sliderSize) : 0);
     }
-
```
Comment (jellysnake) [inline engine/src/main/java/org/terasology/rendering/nui/widgets/UIScrollbar.java]: Only change in file is this deleted line. Again, probably shouldn't be included

File: engine/src/main/java/org/terasology/rendering/nui/widgets/UISlider.java [hunk 11]
```diff
@@ -39,6 +39,8 @@
     public static final String SLIDER = "slider";
     public static final String TICKER = "ticker";
 
+    private int nearestSlot;
+
```
Comment (jellysnake) [inline engine/src/main/java/org/terasology/rendering/nui/widgets/UISlider.java]: Also only contains trivial changes that shouldn't be in this PR

**Comment by jellysnake** [comment]
Found a few more minor bugs.

1. If a checkbox is disabled then you can tab past it fine you just can't tab _back_ using shift. In the image I can tab from the start to the currently selected checkbox but I can't then go backwards to the shadows dropdown. image.
2. It takes two key presses of `tab` before tabbing starts. I don't know if this is intentional but it's consistent. Simply click empty space to deselect and then press tab once. You should notice nothing happens. If you then press it again it will start tabbing. This occurs when paired with shift as well.
3. If you tab to a checkbox and then scroll the scrollable area the checkbox is visually deselected. This is only visual and you can still toggle via `enter`. This also happens when you click it manually so it _may_ not be your code that's causing it.
4. I still can't get checkboxes to work when using `enter`. This applies to both when I tab to select it or when I click to select it.
5. When I'm using the module selection screen, I can select a module and hence activate it. However I can't then deactivate it. Either via the deactivate button or via `enter` again. Additionally it doesn't visually show as activated. (red empasis added by me) 
image. Again, seeing as this also occurs when you select by clicking or pressing select it may not be a result of this PR. This does only occur when you've tabbed at some point in that screen however. I wasn't able to replicate this with further testing nor did I notice anything in the log so I'll just keep an eye out if I can reliably reproduce it.
6. If you tab to select any non-dropdown and then click on a dropdown, [every subsequent keypress gives this in the log](https://hastebin.com/jatepaduxo.txt). It doesn't crash though. This can be reproduced for seemingly any widget, it just changes what it's trying to cast to dropdown
3. _This isn't a bug per se_ but you can tab out of view. It would be nice to have the selected widget always in view.
4. Similarly, _not a bug_ but it would be nice to have tabbing start from the currently selected widget. That seems like it would be more intuitive.
8. _Not a bug_ but would be nice to have UI TextAreas as tab-able along with the sliding selector things.



I'd be happy to merge this with with only fixing item 4, 6 and the outstanding review comments.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by Cervator** [comment]
Tested out - so glad we finally have this. Thank you! :-)

I noticed a few quirks but expect we'll find those for a little while so I'm going to merge this to make them easier to find by more sets of eyes.

* Sliders didn't work for me on the audio menu
* Pressing `ESC` seemed to break further attempts to tab around on a page (due to losing focus somehow?)
* The server status on "Join Server" doesn't seem to refresh? That usually happened when a server entry in the list became selected. Does seem like it gets queried, just no info shows up. Joining seems to still work

Submitted the initial notes in #3577 for further troubleshooting and fixing.

@RatMoleRat again - super appreciated! This has been a long time coming and isn't easy :-)

**Pull request merged**